src/openms/source/CHEMISTRY/EnzymesDB.cpp

```diff
@@ -0,0 +1,331 @@
+  }
+
+  void EnzymesDB::setEnzymes(const String& file_name)
+  {
+    clearEnzymes_();
+    readEnzymesFromFile_(file_name);
+    buildEnzymeNames_();
+  }
+
+  void EnzymesDB::addEnzyme(const Enzyme& enzyme)
+  {
+    Enzyme* r = new Enzyme(enzyme);
+    addEnzyme_(r);
+  }
+
+  void EnzymesDB::addEnzyme_(Enzyme* r)
+  {
+    enzymes_.insert(r);
+    const_enzymes_.insert(r);
+    vector<String> names;
+    if (r->getName() != "")
+    {
+      names.push_back(r->getName());
+    }
+    set<String> synonyms = r->getSynonyms();
+    for (set<String>::iterator it = synonyms.begin(); it != synonyms.end(); ++it)
+    {
+      names.push_back(*it);
+    }
+    buildEnzymeNames_();
+    return;
+  }
+
+  bool EnzymesDB::hasRegEx(const String & cleavage_regex) const
+  {
+    if (enzyme_regex_.has(cleavage_regex))
+    {
+      return true;
+    }
+    return false;
+  }
+
+  bool EnzymesDB::hasEnzyme(const String& enzy_name) const
+  {
+    if (enzyme_names_.find(enzy_name) != enzyme_names_.end())
+    {
+      return true;
+    }
+    return false;
+  }
+  
+  bool EnzymesDB::hasEnzyme(const Enzyme* enzyme) const
+  {
+    if (const_enzymes_.find(enzyme) != const_enzymes_.end() )
+    {
+      return true;
+    }
+    return false;
+  }
+
+  void EnzymesDB::readEnzymesFromFile_(const String& file_name)
+  {
+    String file = File::find(file_name);
+
+    Param param;
+    ParamXMLFile paramFile;
+    paramFile.load(file, param);
+
+    if (!param.begin().getName().hasPrefix("Enzymes"))
+    {
+      throw Exception::ParseError(__FILE__, __LINE__, __PRETTY_FUNCTION__, "", "");
+    }
+
+    try
+    {
+      vector<String> split;
+      param.begin().getName().split(':', split);
+      String prefix = split[0] + split[1];
+      Enzyme* enzy_ptr = 0;
+    
+      Map<String, String> values;
+
+      for (Param::ParamIterator it = param.begin(); it != param.end(); ++it)
+      {
+        it.getName().split(':', split);
+        if (prefix != split[0] + split[1])
+        {
+          // add enzyme
+          enzy_ptr = parseEnzyme_(values);
+          values.clear();
+          enzymes_.insert(enzy_ptr);
+          const_enzymes_.insert(enzy_ptr);
+          prefix = split[0] + split[1];
+        }
+        String value = it->value;
+        String key = it.getName();
+        values[key] = value;
+      }
+
+      // add last enzyme
+      enzy_ptr = parseEnzyme_(values);
+      enzymes_.insert(enzy_ptr);
+      const_enzymes_.insert(enzy_ptr);
+    }
+    catch (Exception::BaseException& e)
+    {
+      throw Exception::ParseError(__FILE__, __LINE__, __PRETTY_FUNCTION__, e.what(), "");
+    }
+  }
+
+  void EnzymesDB::clear_()
+  {
+    clearEnzymes_();
+  }
+
+  void EnzymesDB::clearEnzymes_()
+  {
+    set<Enzyme*>::iterator it;
+    for (it = enzymes_.begin(); it != enzymes_.end(); ++it)
+    {
+      delete *it;
+    }
+    enzymes_.clear();
+    enzyme_names_.clear();
+    enzyme_regex_.clear();
+    const_enzymes_.clear();
```

Comment:
I am sure there is a clever casting macro that could resolve this :-) 

see for example here https://github.com/hroest/OpenMS/blob/feature/replace_DataProcessing/src/openms/include/OpenMS/CONCEPT/Helpers.h#L50 where we have to cast `std::vector<boost::shared_ptr<T> ` to `std::vector<boost::shared_ptr<const T>` why not do the same thing here (without the boost in the middle)?